src/lib/staged_ledger/staged_ledger.ml

```diff
@@ -1983,11 +1983,89 @@ module T = struct
+    type txn =
+      ( Signed_command.With_valid_signature.t
+      , Zkapp_command.Valid.t )
+      User_command.t_
+
+    type t =
+      { valid_seq : txn Sequence.t
+      ; invalid : (txn * Error.t) list
+      ; skipped_by_fee_payer : txn list Account_id.Map.t
+      ; zkapp_space_remaining : int option
+      ; total_space_remaining : int
+      }
+
+    let init ~total_limit ?zkapp_limit =
+      { valid_seq = Sequence.empty
+      ; invalid = []
+      ; skipped_by_fee_payer = Account_id.Map.empty
+      ; zkapp_space_remaining = zkapp_limit
+      ; total_space_remaining = total_limit
+      }
+
+    let txn_key = function
+      | User_command.Zkapp_command cmd ->
+          Zkapp_command.(Valid.forget cmd |> fee_payer)
+      | User_command.Signed_command cmd ->
+          Signed_command.(forget_check cmd |> fee_payer)
+
+    let add_skipped_txn t (txn : txn) =
+      Account_id.Map.update t.skipped_by_fee_payer (txn_key txn)
+        ~f:(Option.value_map ~default:[ txn ] ~f:(List.cons txn))
+
+    let dependency_skipped txn t =
+      Account_id.Map.find t.skipped_by_fee_payer (txn_key txn) |> Option.is_some
+
+    let try_applying_txn ~logger ~validate (state : t) (txn : txn) =
+      let open Continue_or_stop in
+      match txn with
+      | _ when state.total_space_remaining < 1 ->
+          Stop (state.valid_seq, state.invalid)
+      | User_command.Zkapp_command _
+        when Option.value_map ~default:false ~f:(( > ) 1)
+               state.zkapp_space_remaining ->
+          Continue
+            { state with skipped_by_fee_payer = add_skipped_txn state txn }
+      | _ when dependency_skipped txn state ->
+          Continue
+            { state with skipped_by_fee_payer = add_skipped_txn state txn }
+      | _ -> (
+          match
+            O1trace.sync_thread "validate_transaction_against_staged_ledger"
+              (fun () ->
+                validate (Transaction.Command (User_command.forget_check txn)) )
+          with
+          | Error e ->
+              [%log error]
+                ~metadata:
+                  [ ("user_command", User_command.Valid.to_yojson txn)
+                  ; ("error", Error_json.error_to_yojson e)
+                  ]
+                "Staged_ledger_diff creation: Skipping user command: \
+                 $user_command due to error: $error" ;
+              Continue { state with invalid = (txn, e) :: state.invalid }
+          | Ok _txn_partially_applied ->
+              let valid_seq =
+                Sequence.append (Sequence.singleton txn) state.valid_seq
+              in
+              let zkapp_space_remaining =
+                Option.map state.zkapp_space_remaining ~f:(fun limit ->
+                    match txn with
+                    | Zkapp_command _ ->
+                        limit - 1
+                    | Signed_command _ ->
+                        limit )
+              in
+              Continue { state with valid_seq; zkapp_space_remaining } )
```

Comment:
You are not decrementing out `total_space_remaining` here.

Reply:
Indeed.

Reply:
I still don't see a decrement to the total remaining space here after your latest commits. If we don't decrement that, my reading of the code is that we will include as many txns as we can and can generate staged ledger diffs that are greater than the max size.